### DISCUSSION TITLE
Refactor password strength policy to enforce actual strong passwords

### DISCUSSION THREAD

**Discussion by henk23** [directus/directus#17892] [Feature Requests]
### Summary

Currently, when choosing "strong" password policy for a project, these restrictions are applied:

- minimum eight characters
- at least one number
- at least one lowercase letter
- at least one uppercase letter
- at least one of these special characters: `!@#$%^&*()_+}{';'?>.<,`
- no other special characters allowed
- no spaces allowed

This reminded me of the wonderful (but badly-named) project https://github.com/duffn/dumb-password-rules

Rules like this don't really enforce strong passwords, as something easy-to-guess like `Test123!` will pass this policy.
And on the other hand, a good, strong, random password like `0Cxo/CP/PQbweXZwgoTMiwWvQAKy0+9MH4g=` is rejected by Directus with the not very insightful message "Value doesn't have the correct format".

Here are my suggestions:
- Don't restrict what the user can enter into the password field. Let them use spaces, emoji, any characters, including control characters.
- In the project settings
  - instead of choosing none/weak/strong/other policy, just choose the minium password length. The default should be 10 or 12,
  - offer a separate option for enforcing "at least one of number/uppercase/lowercase/specialchar"
  - offer an option for enforcing "password should not contain name, email or other personal details of the user"
  - offer an option to check passwords against the most common 100.000 or so passwords and reject when a match is found.
  - offer an option to check against haveibeenpwned database and reject breached passwords.
- When a password is rejected by any of these policies, clearly communicate to the user what the problem is, which rule they violated and what they can do to find a better password.

### Motivation

My users are frustrated with the rigid password policies and the unhelpful error message.

### Detailed Design

Sorry, just wanted to open a discussion. Have not really thought about implementation details.

### Must Have:
- No restrictions on allowed characters
- Insightful error message

### Should Have:
- Check against common password list
- Check against haveibeenpwned

### Could Have:
- There may be many more good suggestions
- Inspiration could come from here for example: https://www.securden.com/blog/top-10-password-policies.html

### Won't Have:
- ?

### Drawbacks

Implementation-heavy

### Alternatives

Everything can stay as it is, I guess?

### Adoption Strategy

Sorry, don't know Directus well enough, just want to start a discussion.

**Comment by rijkvanzanten** [comment]
Just for some background info on why this is the way it is currently: It's using a RegEx. You could theoretically achieve some of those points by adding a smarter regex under "Other", though things like password checks against databases of existing passwords won't be possible that way of course

**Reply by ruihildt** [reply]
Allowing custom regex would solve the issue then.

It seems to me its an acceptable solution.

**Reply by br41nslug** [reply]
Custom regex is already available however be careful to test any regex used for possible denial of service bugs https://owasp.org/www-community/attacks/Regular_expression_Denial_of_Service_-_ReDoS

**Comment by br41nslug** [comment]
> This reminded me of the wonderful (but badly-named) project https://github.com/duffn/dumb-password-rules

That repo is very aptly named after rules that on first glance seem to make passwords stronger but in reality make them easier to guess and harder to remember (unless password managers are used) 😬Institutions like NIST have also stopped recommending most of these password creating rules in favor of more modern approaches like enforcing 2fa, password audits, checking against breaches

https://blog.netwrix.com/2022/11/14/nist-password-guidelines/
Or the original but thats very dry/technical
https://pages.nist.gov/800-63-3/sp800-63b.html

Having said that for context I do agree we could bring these options more in line with modern recommendations

**Reply by rijkvanzanten** [reply]
I'm almost wondering if we should look into switching to https://github.com/directus/directus/discussions/18755 by default, and only having passwords as a secondary less-secure option 🤔

**Reply by br41nslug** [reply]
Im all for implementing fido/passkeys as a more secure option however im on the fence about defaulting to it that does significantly increase the barrier to entry for trying out the software i think 🤔

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
